File: quiz_engine/test/unit/AnswerShuffler.test.ts
import { shuffleAnswers } from '../../src/service/AnswerShuffler';
import { Question } from '../../src/models/Question';

function makeQuestion(overrides: Partial<Question> = {}): Question {
  const q = new Question();
  q.id = 1;
  q.questionText = 'Test question?';
  q.optionA = 'Alpha';
  q.optionB = 'Beta';
  q.optionC = 'Gamma';
  q.optionD = 'Delta';
  q.correctAnswer = 'A';
  q.usageCycle = 1;
  q.timesUsed = 0;
  q.createdAt = new Date();
  Object.assign(q, overrides);
  return q;
}

describe('AnswerShuffler', () => {
  it('returns all original options after shuffling', () => {
    const q = makeQuestion();
    const result = shuffleAnswers(q);
    expect(new Set(result.shuffledOptions)).toEqual(
      new Set([q.optionA, q.optionB, q.optionC, q.optionD]),
    );
  });

  it('always returns 4 options for a 4-option question', () => {
    const q = makeQuestion();
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions).toHaveLength(4);
  });

  it('returns 5 options for a 5-option question', () => {
    const q = makeQuestion({ optionE: 'Epsilon' });
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions).toHaveLength(5);
    expect(result.shuffledOptions).toContain('Epsilon');
  });

  it('maps correctShuffledIndex to the correct answer text', () => {
    const q = makeQuestion({ correctAnswer: 'A' }); // A = Alpha
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Alpha');
  });

  it('maps correct answer B correctly', () => {
    const q = makeQuestion({ correctAnswer: 'B' }); // B = Beta
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Beta');
  });

  it('maps correct answer C correctly', () => {
    const q = makeQuestion({ correctAnswer: 'C' }); // C = Gamma
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Gamma');
  });

  it('maps correct answer D correctly', () => {
    const q = makeQuestion({ correctAnswer: 'D' }); // D = Delta
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Delta');
  });

  it('maps correct answer E correctly', () => {
    const q = makeQuestion({ optionE: 'Epsilon', correctAnswer: 'E' });
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Epsilon');
  });

  it('falls back to optionD when E is correct but optionE missing', () => {
    const q = makeQuestion({ correctAnswer: 'E' }); // No optionE
    const result = shuffleAnswers(q);
    // Should not throw and should return a valid index
    expect(result.correctShuffledIndex).toBeGreaterThanOrEqual(0);
  });

  it('displayToOriginal has an entry for each display letter', () => {
    const q = makeQuestion();
    const result = shuffleAnswers(q);
    expect(Object.keys(result.displayToOriginal)).toHaveLength(4);
  });

  it('displayToOriginal maps display letters to original letters', () => {
    const q = makeQuestion();
    const result = shuffleAnswers(q);
    const letters = ['A', 'B', 'C', 'D'];
    for (const letter of letters) {
      expect(letters).toContain(result.displayToOriginal[letter]);
    }
  });

  it('is non-deterministic (shuffles differently at least sometimes)', () => {
    // Run many times and check that not all are identical
    const q = makeQuestion();
    const indices = new Set<number>();
    for (let i = 0; i < 50; i++) {
      indices.add(shuffleAnswers(q).correctShuffledIndex);
    }
    // With 4 positions, over 50 runs we expect more than 1 unique index
    expect(indices.size).toBeGreaterThan(1);
  });

  it('handles unknown correct answer letter gracefully', () => {
    const q = makeQuestion({ correctAnswer: 'Z' }); // Falls back to A
    const result = shuffleAnswers(q);
    expect(result.shuffledOptions[result.correctShuffledIndex]).toBe('Alpha');
  });
});

File: quiz-engine/quiz-engine-dart/test/src/service/answer_shuffler_test.dart
import 'package:test/test.dart';
import 'package:quiz_engine_dart/src/service/answer_shuffler.dart';
import 'package:quiz_engine_dart/src/models/question.dart';

Question _makeQuestion({
  String optionA = 'Alpha',
  String optionB = 'Beta',
  String optionC = 'Gamma',
  String optionD = 'Delta',
  String? optionE,
  String correctAnswer = 'A',
}) {
  return Question(
    id: 1,
    questionText: 'Test?',
    optionA: optionA,
    optionB: optionB,
    optionC: optionC,
    optionD: optionD,
    optionE: optionE,
    correctAnswer: correctAnswer,
    createdAt: DateTime.now(),
  );
}

void main() {
  late AnswerShuffler shuffler;

  setUp(() => shuffler = AnswerShuffler());

  group('shuffle', () {
    test('preserves all original options', () {
      final q = _makeQuestion();
      final result = shuffler.shuffle(q);
      final expected = {'Alpha', 'Beta', 'Gamma', 'Delta'};
      expect(result.shuffledOptions.toSet(), equals(expected));
    });

    test('returns correct number of options (4)', () {
      final q = _makeQuestion();
      final result = shuffler.shuffle(q);
      expect(result.shuffledOptions.length, equals(4));
    });

    test('returns 5 options when optionE is present', () {
      final q = _makeQuestion(optionE: 'Epsilon');
      final result = shuffler.shuffle(q);
      expect(result.shuffledOptions.length, equals(5));
    });

    test('maps correct answer to new shuffled position', () {
      final q = _makeQuestion(correctAnswer: 'A');
      final result = shuffler.shuffle(q);
      final correctIndex =
          result.labels.indexOf(result.correctLabel);
      expect(result.shuffledOptions[correctIndex], equals('Alpha'));
    });

    test('correctLabel is one of the valid labels', () {
      final q = _makeQuestion();
      final result = shuffler.shuffle(q);
      expect(result.labels, contains(result.correctLabel));
    });

    test('labels are A–D for 4 options', () {
      final q = _makeQuestion();
      final result = shuffler.shuffle(q);
      expect(result.labels, equals(['A', 'B', 'C', 'D']));
    });

    test('labels are A–E for 5 options', () {
      final q = _makeQuestion(optionE: 'Epsilon');
      final result = shuffler.shuffle(q);
      expect(result.labels, equals(['A', 'B', 'C', 'D', 'E']));
    });

    test('shuffles answer B correctly', () {
      final q = _makeQuestion(correctAnswer: 'B');
      final result = shuffler.shuffle(q);
      final correctIndex = result.labels.indexOf(result.correctLabel);
      expect(result.shuffledOptions[correctIndex], equals('Beta'));
    });

    test('shuffles answer C correctly', () {
      final q = _makeQuestion(correctAnswer: 'C');
      final result = shuffler.shuffle(q);
      final correctIndex = result.labels.indexOf(result.correctLabel);
      expect(result.shuffledOptions[correctIndex], equals('Gamma'));
    });

    test('shuffles answer D correctly', () {
      final q = _makeQuestion(correctAnswer: 'D');
      final result = shuffler.shuffle(q);
      final correctIndex = result.labels.indexOf(result.correctLabel);
      expect(result.shuffledOptions[correctIndex], equals('Delta'));
    });

    test('shuffles answer E correctly for 5-option question', () {
      final q = _makeQuestion(optionE: 'Epsilon', correctAnswer: 'E');
      final result = shuffler.shuffle(q);
      final correctIndex = result.labels.indexOf(result.correctLabel);
      expect(result.shuffledOptions[correctIndex], equals('Epsilon'));
    });
  });
}
